networklibrary/src/main/java/com/eis/smsnetwork/SMSNetSubscriberList.java

```diff
@@ -6,16 +6,18 @@
 import com.eis.communication.network.NetSubscriberList;
 import com.eis.smslibrary.SMSPeer;
 
-import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Set;
 
 /**
  * Concrete implementation of a {@link NetSubscriberList} interface
  *
  * @author Marco Cognolato
+ * @author Giovanni Velludo
  */
 public class SMSNetSubscriberList implements NetSubscriberList<SMSPeer> {
```

Comment:
This entire class could be replaced by a `Set<SMSPeer>` in SMSNetworkManager

Reply:
Could somebody share their opinion? I'm asking here instead of doing the change myself because this code was already submitted for a review.

Reply:
These are the differences between a Set and an ArrayList
![image](https://user-images.githubusercontent.com/22344699/74137822-c6eee280-4bf0-11ea-89c6-acdef7e5ea39.png)

Given these differences, I guess SMSNetSubscriberList could be both a Set or an ArrayList.
Yeah a set cannot have duplicates, but the SubscriberList shouldn't have duplicates in the first place, so we can solve it in the ArrayList by just checking before hand.
The reason I would keep the ArrayList is because you can have ordered elements, while with the Set it's open to implementation, and I don't want to deal with that. Having a ordered array and considering disordered is easier than having a set and needing to have it ordered

Reply:
We don't need the list of subscribers to be ordered, and a HashSet has better performance than an ArrayList for searches, which we'll do when checking if the list contains a certain number, and also when removing numbers from the list.

Reply:
> We don't need the list of subscribers to be ordered

yeah, WE don't need it.
But I remember you that there's a method to replace the SubscriberList from outside, so, what if someone needs the list to be ordered?
Also group 4 actually DOES need it to be ordered, so they DO use that method to replace the subscriber list, just so they can have an ordered list.

> a HashSet has better performance than an ArrayList for searches

Yeah, you're totally right, but... do we need to do searches?
We definitely need if it's an arraylist, but with the Set, do we?
Well no, because since the set cannot have duplicates, you don't need to check if there's already a peer when you're adding a new one, and when you remove a peer from the list, you don't neet to check if there is either, you'll just get returned null if there's not, but who cares, you just need to remove it, no one will know that it wasn't there in the first place

Reply:
> But I remember you that there's a method to replace the SubscriberList from outside, so, what if someone needs the list to be ordered?
Also group 4 actually DOES need it to be ordered, so they DO use that method to replace the subscriber list, just so they can have an ordered list.

When did they tell us that they need the list of subscribers to be ordered? I can't find any mention of that [in the interfaces they gave us](https://github.com/Cogno-IDU/network-dictionary/pull/6/files), according to which we developed the replicated network.

> Yeah, you're totally right, but... do we need to do searches?
We definitely need if it's an arraylist, but with the Set, do we?

We run a search on the list of subscribers when checking if a message we received came from a subscriber, which is when we call `contains()` at line 70 in [BroadcastReceiver](https://github.com/Cogno-IDU/network-dictionary/pull/9/files#diff-d040282b44fdb3c2f637d3ca5c1cb471). Method `contains()` searches for the given `SMSPeer` in the `ArrayList`, and it would be more efficient on a `HashSet`.
We also perform a search when removing a `SMSPeer` from the list of subscribers, because `contains()` is called by method `remove()` to check if the `SMSPeer` to remove is present in the list.

> Well no, because since the set cannot have duplicates, you don't need to check if there's already a peer when you're adding a new one, and when you remove a peer from the list, you don't neet to check if there is either, you'll just get returned null if there's not, but who cares, you just need to remove it, no one will know that it wasn't there in the first place

We don't do those searches in our code, but they're done by `ArrayList` and `HashSet`'s implementation of methods `contains()` and `remove()`. `ArrayList` doesn't search for a peer when adding it, because it allows duplicates. This is bad because if we allow duplicates to exist, we won't remove all copies of the duplicate when calling `ArrayList.remove()`.

Reply:
> When did they tell us that they need the list of subscribers to be ordered?

It should be in pr #8

Reply:
I can't find it, could you link the post where they requested it?

Reply:
Woops, it was in pr #7
[here it is](https://github.com/Cogno-IDU/network-dictionary/pull/7#discussion_r365596463)

Reply:
Ok, so instead of a `Set<SMSPeer>` in `SMSNetworkManager` we could use a `Collection<SMSPeer>`, so that it can be replaced with an ordered `ArrayList` or whatever data structure the user of this library wants:
```java
private Collection<SMSPeer> netSubscribers = new HashSet<>();
```
Because the way we're using `NetSubscriberList` is as a `Collection`, so it seems like a pointless class.

Reply:
Well, if you replace with a Collection interface, now the user who wants to add a simple implementation of a list has to implement all of [these methods](https://docs.oracle.com/javase/8/docs/api/java/util/Collection.html).
If we keep this interface/implementation the user only has to implement those, which are also the only one we use.
Even though you can just add an ArrayList from outside, you can't  choose how it behaves unless you implement the interface in a new class and code it yourself.

Reply:
I thought realistically they would probably use one of the data structures already provided by the Java SE API, but you're right, `NetSubscribersList` is easier to implement from scratch.

Reply:
If you need to replace that is because you want more control, if you want more control you have to create the behaviour yourself, using the interface becomes extra work for no reason.
I'll mark this as resolved then.